File: solutions/scala/collatz-conjecture/src/main/scala/CollatzConjecture.scala
object CollatzConjecture:

  def steps(number: Int): Option[Int] =
    @scala.annotation.tailrec
    def loop(number: Int, step: Int): Int =
      number match
        case 1 => step
        case x if x % 2 == 0 => loop(x / 2, step + 1)
        case x => loop(3 * x + 1, step + 1)

    Option.when(number > 0):
      loop(number, 0)

File: solutions/fsharp/collatz-conjecture/CollatzConjecture.fs
module CollatzConjecture

let rec count number : int =
    if number < 1 then 1
    elif number % 2 = 0 then 1 + count number / 2
    else 1 + count 3 * number + 1
let rec steps (number: int): int option =
    match number with
    | _ when number < 0 -> None
    | 1 -> Some 1
    | _ -> Some <| count number
    
    
     